### ISSUE TITLE
Build failure on Cabal and Just grammars

### ISSUE THREAD

**Issue by heliostatic** [helix-editor/helix#6788]
### Summary

Getting build failures with the new Cabal (#6485) grammar and Just. Appears to be C++ related?

### Reproduction Steps

 

I tried this:

1. `hx --grammar build`

I expected this to happen:

Grammars built with no errors

Instead, this happened:

Errors building `just` and `cabal` grammars.

<details><summary>grammar build log</summary>

```
Building 149 grammars
147 grammars already built
2 grammars failed to build
        Failure 0/2: Parser compilation failed.
Stdout:
Stderr: /Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:15:24: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
  uint32_t prev_indent = 0;
                       ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:47:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto curr = std::to_string(state->prev_indent);
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:64:14: error: expected expression
    *state = {};
             ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:88:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto advance = [lexer] { lexer->advance(lexer, false); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:88:18: error: expected expression
  auto advance = [lexer] { lexer->advance(lexer, false); };
                 ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:89:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto skip = [lexer] { lexer->advance(lexer, true); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:89:15: error: expected expression
  auto skip = [lexer] { lexer->advance(lexer, true); };
              ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:91:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto get_column = [lexer] { return lexer->get_column(lexer); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:91:21: error: expected expression
  auto get_column = [lexer] { return lexer->get_column(lexer); };
                    ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:128:5: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
    auto indent = get_column();
    ^
6 warnings and 4 errors generated.

        Failure 1/2: Parser compilation failed.
Stdout:
Stderr: /Users/bc/.config/helix/runtime/grammars/sources/cabal/src/scanner.cc:26:60: error: expected expression
        for_each(indent_lvls.cbegin(), indent_lvls.cend(), [buffer, &i](uint8_t const n) { buffer[i++] = n; });
                                                           ^
/Users/bc/.config/helix/runtime/grammars/sources/cabal/src/scanner.cc:70:9: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto cur_indent_lvl = indent_lvls.back();
        ^
1 warning and 1 error generated.
```
</details>

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
```

</details>

### Platform

macOS

### Terminal Emulator

Alacritty

### Helix Version

helix 23.03 (4cdba7cc)

**Comment by pascalkuthe** [comment]
seems like your c++ compiler might be very old and doesn't use c++11 by default. You can try `c++ --version` to figure out your compiler version.  Macos ships with clang which defaults to c++14 since version 6.0 (which is 5 years old)

**Comment by pascalkuthe** [comment]
Maybe apple-clang has a different default? Googling seems to turn up other people with a different problem, but other people always claim it works. To enforce a newer standard we could pass `-std=c++14` manually but that's not quite portable (because windows does flags differently...)  sadly the `cc` crate doesn't have builtin functionality for that but we could use the workaround suggested here https://github.com/rust-lang/cc-rs/issues/565#issuecomment-1336416866 for now.

**Comment by pascalkuthe** [comment]
that shouldn't matter we invoke the c compiler directly tough the CC crate. `binding.gyp` is for integration with `node.js` and not used by helix in any way

**Comment by h1t** [comment]
We can't compile C and Cpp files in one command using `-std=c99` for C and `-std=c++11` for Cpp.
We have to do it with two separate commands.

**Comment by h1t** [comment]
clang 14 on macos
```
$ clang -xc ./parser.c -std=c++11 ./scanner.cc
error: invalid argument '-std=c++11' not allowed with 'C'
```

**Comment by pascalkuthe** [comment]
To use `-std=c++14` you need to use the `clang++`. They are both the same binary but the `clang++` driver behaves differently. #6792  should fix the problem. Please try building with that PR @heliostatic

**Comment by pascalkuthe** [comment]
Ok this was interesting, https://github.com/helix-editor/helix/pull/6792 fixed this problem and with  #6795 this actually fails in CI too (it already does right now) as we were simply ignoring grammar build failures by accident in the past.

With those two PRs the problem is fixed and can be automatically tested in CI (so similar problems don't reappear in the future)
